Frameworks/UIKit/UIViewController.mm

```diff
@@ -1211,57 +1211,71 @@ - (void)_dismissTransitionStopped:(id)anim finished:(BOOL)finished {
     [[UIApplication sharedApplication] endIgnoringInteractionEvents];
 }
 
+- (void)_childDismissCleanup {
+    priv->_presentedViewController->priv->_parentViewController = nil;
+    priv->_presentedViewController->priv->_presentingViewController = nil;
+    priv->_presentedViewController->priv->_popoverPresentationController = nil;
+    priv->_presentedViewController = nil;
+}
+
 /**
  @Status Interoperable
 */
 - (void)dismissViewControllerAnimated:(BOOL)animated completion:(void (^)(void))completion {
-    if (priv->_modalViewController == nil) {
-        if ([self parentViewController] != nil) {
+    if (![self presentedViewController]) {
+        // Calling dismiss on a view controller that hasn't presented additional view controllers itself should result in the parent handling the dismissal.
+
+        if ([self parentViewController]) {
             [[self parentViewController] dismissViewControllerAnimated:animated completion:completion];
-            return;
+        } else {
+            TraceWarning(TAG, L"dismissViewController invalid - nothing to dismiss");
         }
 
-        TraceWarning(TAG, L"dismissModalViewController invalid!");
         return;
     }
 
-    UIViewController* curController = priv->_modalViewController;
+    UIViewController* presented = [self presentedViewController];
+
+    if ([presented presentedViewController]) {
+        // Recursively handle multiple child dismissal.
+        // TODO: This is unsafe because we end up dispatching the dimiss animation of a
+        // parent before the child dispatch animation completes.
+        [presented dismissViewControllerAnimated:animated completion:completion];
+    }
 
-    [curController retain];
-    [curController autorelease];
+    __block __unsafe_unretained UIViewController *weakSelf = self;
+    dispatch_block_t cleanupCompletion = ^{
+        StrongId<UIViewController> strongSelf = weakSelf;
+        [strongSelf _childDismissCleanup];
+
+        // Completion must happen after full tear-down, otherwise attempts to present a new
+        // view controller from the parent of the just dismissed controller, in this completion,
+        // will fail on the stale presentingViewController/presentedViewController relationship.
+        if (completion) {
+            completion();
+        }
+    };
 
     // We maintain a popoverPresentationController instance whenever UIModalPresentationPopover is specified.
     // However, in non-tablet operation mode, UIModalPresentationPopover should result in the presentation of
     // a full screen modal instead. We check here if an actual popover has been presented to handle the
     // dismiss (of a genuine popover or alternately a full screen modal) appropriately.
```

Comment:
I think it makes sense to handle removing the tablet vs nontablet logic now since it affects presentViewController. The pull in #1511 includes this but I think would be better to bring in the additional changes as a third commit in this pull for clarity, then close #1511.